### PR TITLE
Eliminate more redundant copies

### PR DESCRIPTION
**Pull request by slavapestov**
Eliminate another source of redundant copies in protocol witness thunks, and clean up some unrelated code in the AST.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please test

File: lib/SILGen/Scope.cpp [hunk 1]
```diff
@@ -119,31 +119,6 @@ RValue Scope::popPreservingValue(RValue &&rv) {
   return RValue(SGF, std::move(managedValues), type, numEltsRemaining);
 }
 
-void Scope::popPreservingValues(ArrayRef<ManagedValue> innerValues,
```
Comment (gottesmm) [inline lib/SILGen/Scope.cpp]: Why are you removing this, this is useful.

**Review by gottesmm — COMMENTED** [review]
Beyond that 1 comment. +1!

File: lib/SILGen/Scope.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/SILGen/Scope.cpp]: I wasn't sure if it was generally useful or a hack before we had tuple restructuring. Also we generally delete utility functions that are not used. However I can bring it back if you want. I'll let the benchmarks finish first then force push the other three patches leaving the one that removes this one behind.

Comment (gottesmm) [reply lib/SILGen/Scope.cpp]: My suggestion: just revert the patch in a later PR. Then your other patches won't be delayed and you can continue with your day = ) (the full tests take a bit longer).

The reason why I am for keeping this is that from an API perspective, it is important for Scope to have parity on this in between RValue and ManagedValue. I do not want people to have to create RValues just to push a value through a scope.

**Pull request merged**